p2p/README.md

```diff
@@ -67,105 +66,431 @@ flowchart TD
 
-subgraph rDHT[Kademlia DHT]
-end
+Messages MUST be deduplicated before broadcasting their respective event over the bus since it is expected that nodes will receive duplicate messages (for multiple reasons).
 
-rGossipSub --> rBGPS
-rDHT --> rBGPS
-end
+The responsibility of deduplication is encapsulated by the P2P module, As such duplicate messages may come from multiple routers in some of these scenarios.
 
-rHost -. "3 (setStreamHandler())" .-> hs[[handleStream]]
+```mermaid
+classDiagram
+    class RainTreeMessage {
+        <<protobuf>>
+        +Level uint32
+        +Data []byte
+    }
+
+    class BackgroundMessage {
+        <<protobuf>>
+        +Data []byte
+    }
+    
+    class PocketEnvelope {
+        <<protobuf>>
+        +Content *anypb.Any
+        +Nonce uint64
+    }
+
+    RainTreeMessage --* PocketEnvelope : serialized as `Data`
+    BackgroundMessage --* PocketEnvelope : serialized as `Data`
+    
+    
+    class p2pModule {
+        -handlePocketEnvelope([]byte) error
+    }
+
+    class P2PModule {
+        <<interface>>
+        GetAddress() (Address, error)
+        HandleEvent(*anypb.Any) error
+        Send([]byte, address Address) error
+        Broadcast([]byte) error
+    }
+    p2pModule --|> P2PModule
+
+    class RainTreeRouter {
+        UnicastRouter
+        -handler MessageHandler
+        +Broadcast([]byte) error
+        -handleRainTreeMsg([]byte) error
+    }
+
+    class NonceDeduper {
+        Push(Nonce) error
+        Contains(Nonce) bool
+    }
+
+  class Bus {
+    <<interface>>
+    PublishEventToBus(PocketEnvelope)
+    GetBusEvent() PocketEnvelope
+  }
+  p2pModule --> Bus
+
+    class BackgroundRouter {
+        UnicastRouter
+        -handler MessageHandler
+        +Broadcast([]byte) error
+        -handleBackgroundMsg([]byte) error
+        -readSubscription(subscription *pubsub.Subscription)
+    }
+
+    class UnicastRouter {
+        -messageHandler MessageHandler
+        -peerHandler PeerHandler
+        +Send([]byte, address Address) error
+        -handleStream(stream libp2pNetwork.Stream)
+        -readStream(stream libp2pNetwork.Stream)
+    }
+    RainTreeRouter --* UnicastRouter : (embedded)
+    BackgroundRouter --* UnicastRouter : (embedded)
+
+    p2pModule ..* RainTreeRouter
+    RainTreeRouter --o RainTreeMessage
+    
+    p2pModule ..* BackgroundRouter
+    BackgroundRouter --o BackgroundMessage
+
+    p2pModule --o PocketEnvelope
+    p2pModule --* NonceDeduper
+```
 
-hs --4a--> rRT
-hs --4b--> rBG
-rBG  --"5a (cont. propagation)"--> rHost
-linkStyle 11 stroke:#ff3
-rRT  --"5b (cont. propagation)"--> rHost
-linkStyle 12 stroke:#ff3
-end
+### Peer Discovery
+Peer discovery involves pairing peer IDs to their network addresses (multiaddr).
+This pairing always has an associated TTL (time-to-live), near the end of which it must
+be refreshed.
+
+In the background gossip overlay network (`backgroundRouter`), peers will re-advertise themselves 7/8th through their TTL.
```

Comment:
What do you mean by 7/8th?

Reply:
Thanks for calling this out. I intended to provide a reference here.

When I wrote that I was thinking of [this discovery util](https://github.com/libp2p/go-libp2p/blob/v0.28.1/p2p/discovery/util/util.go#L50) which would mean literally 7/8 = 0.875 = 87.5% of the way through their TTL. However, upon further inspection, I don't think this is actually what's being used in the context of the DHT. It looks like  peers which are "DHT providers" (currently all by default, I think) get a 24 hour TTL by default but they are expected to re-advertise every 3 hours.

```suggestion
In the background gossip overlay network (`backgroundRouter`), peers will re-advertise themselves every 3 hours through their TTL (see: [`RoutingDiscovery#Advertise()`](https://github.com/libp2p/go-libp2p/blob/87c2561238cb0340ddb182c61be8dbbc7a12a780/p2p/discovery/routing/routing.go#L34) and [`ProviderManager#AddProvider()`](https://github.com/libp2p/go-libp2p-kad-dht/blob/v0.24.2/providers/providers_manager.go#L255)).
```

Reply:
LGTM, thanks for the update!

Reply:
@bryanchriswhite Can you please apply these changes?